### PR TITLE
feat(language-service): Introduce 'angularOnly' flag

### PR DESCRIPTION
**Pull request by kyliau**
This PR changes the language service to work in two different modes:

1. TS + Angular
   Plugin augments TS language service to provide additonal Angular
   information. This only works with inline template and is meant to be
   used as a local plugin (configured via tsconfig.json).
2. Angular only
   Plugin only provides information on Angular templates, no TS info at
   all. This effectively disables native TS features and is meant for
   internal use only.

Default mode is `angularOnly = false` so that we don't break any users
already using Angular LS as local plugin.

As part of the refactoring, `undefined` is removed from type aliases
because it is considered bad practice.

go/tsstyle#nullableundefined-type-aliases
```
Type aliases must not include |null or |undefined in a union type.
Nullable aliases typically indicate that null values are being passed
around through too many layers of an application, and this clouds the
source of the original issue that resulted in null. They also make it
unclear when specific values on a class or interface might be absent.
```

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/language-service/src/language_service.ts [hunk 1]
```diff
@@ -112,25 +112,20 @@ class LanguageServiceImpl implements LanguageService {
...
-      }
-      if (!set.has(span.end)) {
-        set.add(span.end);
-        result.push(element);
-      }
+function uniqueBySpan<T extends{span: Span}>(elements: T[]): T[] {
+  const result: T[] = [];
+  const map = new Map<number, Set<number>>();
+  for (const element of elements) {
+    let span = element.span;
```
Comment (ayazhafiz) [inline packages/language-service/src/language_service.ts]:
```suggestion
    const span = element.span;
```

File: packages/language-service/src/ts_plugin.ts [hunk 2]
```diff
@@ -61,154 +61,158 @@ function diagnosticToDiagnostic(d: Diagnostic, file: ts.SourceFile): ts.Diagnost
...
+      const tsResults = tsLS.getSemanticDiagnostics(fileName);
+      results.push(...tsResults);
+    }
+    // For semantic diagnostics we need to combine both TS + Angular results
+    const ngResults = ngLS.getDiagnostics(fileName);
+    if (!ngResults.length) {
+      return results;
+    }
+    const sourceFile =
+        fileName.endsWith('.ts') ? tsLS.getProgram() !.getSourceFile(fileName) : undefined;
```
Comment (ayazhafiz) [inline packages/language-service/src/ts_plugin.ts]: prefer sanity check that `tsLs.getProgram()` exists

File: packages/language-service/src/ts_plugin.ts [hunk 3]
```diff
@@ -61,154 +61,158 @@ function diagnosticToDiagnostic(d: Diagnostic, file: ts.SourceFile): ts.Diagnost
...
+          return;
+        }
+        return results.map((loc: Location) => {
+          return {
+            fileName: loc.fileName,
+            textSpan: {
+              start: loc.span.start,
+              length: loc.span.end - loc.span.start,
+            },
+            name: '',
```
Comment (ayazhafiz) [inline packages/language-service/src/ts_plugin.ts]: Why is the name empty? What is name supposed to be?

File: packages/language-service/src/ts_plugin.ts [hunk 4]
```diff
@@ -61,154 +61,158 @@ function diagnosticToDiagnostic(d: Diagnostic, file: ts.SourceFile): ts.Diagnost
...
+        const results = ngLS.getDefinitionAt(fileName, position);
+        if (!results || !results.length) {
+          return;
         }
-        const ours = ls.getDefinitionAt(fileName, position);
-        if (ours && ours.length) {
-          return ours.map((loc: Location) => {
+        const {span} = results[0];
+        return {
+          definitions: results.map((loc: Location) => {
```
Comment (ayazhafiz) [inline packages/language-service/src/ts_plugin.ts]: nit: pull this out as a variable

File: packages/language-service/src/types.ts [hunk 5]
```diff
@@ -245,7 +245,7 @@ export interface Completion {
  *
  * @publicApi
  */
-export type Completions = Completion[] | undefined;
+export type Completions = Completion[];
```
Comment (ayazhafiz) [inline packages/language-service/src/types.ts]: Could we get rid of this entirely and use `Completion[]` in all usages instead?

File: packages/language-service/test/ts_plugin_spec.ts [hunk 6]
```diff
@@ -197,8 +183,53 @@ describe('plugin', () => {
...
     });
+
+    describe(`with config 'angularOnly = true`, () => {
+      const ngLS = createPlugin(service, mockHost, {angularOnly: true});
+      it('should not report template errors on TOH', () => {
+        const sourceFiles = ngLS.getProgram() !.getSourceFiles();
+        expect(sourceFiles.length).toBeGreaterThan(0);
+        for (const {fileName} of sourceFiles) {
+          // Ignore all 'cases.ts' files as they intentionally contain errors.
+          if (!fileName.endsWith('cases.ts')) {
```
Comment (ayazhafiz) [inline packages/language-service/test/ts_plugin_spec.ts]: future: maybe rename these to end with `error.ts`

File: packages/language-service/src/language_service.ts (hunk 1 above)
Comment (kyliau) [reply packages/language-service/src/language_service.ts]: done

File: packages/language-service/src/ts_plugin.ts (hunk 2 above)
Comment (kyliau) [reply packages/language-service/src/ts_plugin.ts]: used ngLSHost.getSourceFile() instead

File: packages/language-service/src/ts_plugin.ts (hunk 3 above)
Comment (kyliau) [reply packages/language-service/src/ts_plugin.ts]: should be name of symbol but we don't have enough information here.
Added TODO to address this in future PR.

File: packages/language-service/src/ts_plugin.ts (hunk 4 above)
Comment (kyliau) [reply packages/language-service/src/ts_plugin.ts]: thank you! This is much cleaner after refactoring.

File: packages/language-service/src/types.ts (hunk 5 above)
Comment (kyliau) [reply packages/language-service/src/types.ts]: It's technically a breaking change so I didn't want to remove it entirely. I've marked both `Completions` and `Definitions` as deprecated.

File: packages/language-service/test/ts_plugin_spec.ts (hunk 6 above)
Comment (kyliau) [reply packages/language-service/test/ts_plugin_spec.ts]: added TODO.

**Review by ayazhafiz — APPROVED** [review]

**Pull request closed without merging**